src/main/java/one/util/streamex/DoubleWithFirstSpliterator.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2015, 2017 StreamEx contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package one.util.streamex;
+
+import java.util.Spliterator;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.DoubleBinaryOperator;
+import java.util.function.DoubleConsumer;
+import one.util.streamex.StreamExInternals.DoubleCloneableSpliterator;
+
+
+/**
+ * @author Tagir Valeev
+ */
+/* package */final class DoubleWithFirstSpliterator extends DoubleCloneableSpliterator<DoubleWithFirstSpliterator> implements DoubleConsumer, Spliterator.OfDouble {
+    private static final int STATE_NONE = 0;
+    private static final int STATE_FIRST_READ = 1;
+    private static final int STATE_INIT = 2;
+    private static final int STATE_EMPTY = 3;
+
+    private ReentrantLock lock;
+    private Spliterator.OfDouble source;
+    private DoubleWithFirstSpliterator prefix;
+    private volatile double first;
+    private volatile int state = STATE_NONE;
+    private final DoubleBinaryOperator mapper;
+    private DoubleConsumer action;
+    
+    DoubleWithFirstSpliterator(Spliterator.OfDouble source, DoubleBinaryOperator mapper) {
+        this.source = source;
+        this.mapper = mapper;
+    }
+    
+    private void acquire() {
+        if(lock != null && state == STATE_NONE) {
+            lock.lock();
+        }
+    }
+    
+    private void release() {
+        if(lock != null && lock.isHeldByCurrentThread()) {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public boolean tryAdvance(DoubleConsumer action) {
+        if (state == STATE_NONE) {
+            acquire();
+            try {
+                doInit();
+            }
+            finally {
+                release();
+            }
+        }
+        if (state == STATE_FIRST_READ) {
+            state = STATE_INIT;
+            action.accept(mapper.applyAsDouble(first, first));
+            return true;
+        }
+        if (state != STATE_INIT)
+            return false;
+        this.action = action;
+        boolean hasNext = source.tryAdvance(this);
+        this.action = null;
+        return hasNext;
+    }
+
+    private void doInit() {
+        double prefixState = state;
+        if (prefixState != STATE_NONE)
+            return;
+        if (prefix != null) {
+            prefix.doInit();

```

Comment:
doInit() could be an abstract method of supertype, so prefix.doInit() could be called